### PR TITLE
store: add support for tracks in channels.

### PR DESCRIPTION
**Pull request by josepht**
* since snapcraft pushes channels as strings support
  already exists for pushing/releasing to LTS Channels.

* this commit adds the Arch, Track, and Series to the
  outputted channel map (now channel map tree) as
  returned by the store.

LP:#1663014

### CODE REVIEW

**Comment by sergiusens** [comment]
can you add some live output please?

**Comment by josepht** [comment]
josepht@maple:~/snaps/joe$ ~/src/git/snapcraft/bin/snapcraft push hello-joe_0.2_amd64.snap --release edge
Pushing 'hello-joe_0.2_amd64.snap' to the store.
Uploading hello-joe_0.2_amd64.snap [                                                                                                                                                                                                ]   0%
Uploading hello-joe_0.2_amd64.snap [================================================================================================================================================================================================] 100%
Error while processing...|                                                                                                                                                                                                                
Uploaded 'hello-joe'
There has been a problem while analyzing the snap, check the snap and try to push again.
  - binary_sha3_384: A file with this exact same content has already been uploaded

**Comment by josepht** [comment]
push:

```
josepht@maple:~/snaps/joe$ ~/src/git/snapcraft/tmp/snapcraft/bin/snapcraft push hello-joe_0.3_amd64.snap --release 0.2/edge
Pushing 'hello-joe_0.3_amd64.snap' to the store.
Uploading hello-joe_0.3_amd64.snap [                                                                                                                                                                                                ]   0%
Uploading hello-joe_0.3_amd64.snap [================================================================================================================================================================================================] 100%
Ready to release!|                                                                                                                                                                                                                        
Revision 3 of 'hello-joe' created.
Arch    Track    Series    Channel    Version    Revision
amd64   0.2      16        stable     -          -
                           candidate  -          -
                           beta       -          -
                           edge       0.3        3
```

release:

```
josepht@maple:~/snaps/joe$ ~/src/git/snapcraft/tmp/snapcraft/bin/snapcraft release hello-joe 3 0.2/beta
Arch    Track    Series    Channel    Version    Revision
amd64   0.2      16        stable     -          -
                           candidate  -          -
                           beta       0.3        3
                           edge       0.3        3
The '0.2/beta' channel is now open.
```

**Comment by sergiusens** [comment]
How about a test with more than one arch?

**Comment by josepht** [comment]
Here's a multiarch snap:

```
josepht@maple:~/snaps/joe$ ~/src/git/snapcraft/bin/snapcraft push hello-joe_0.3_multi.snap --release edge           
Pushing 'hello-joe_0.3_multi.snap' to the store.
Uploading hello-joe_0.3_multi.snap [                                                                          ]   0%
Uploading hello-joe_0.3_multi.snap [==========================================================================] 100%
Ready to release!|                                                                                                  
Revision 4 of 'hello-joe' created.
Arch    Track    Series    Channel    Version    Revision
armhf   latest   16        stable     -          -
                           candidate  -          -
                           beta       -          -
                           edge       0.3        4
Arch    Track    Series    Channel    Version    Revision
amd64   latest   16        stable     -          -
                           candidate  -          -
                           beta       0.2        2
                           edge       0.3        4
The 'edge' channel is now open.
```

Once I've built an arm version I'll add that as well.

**Comment by josepht** [comment]
Here's an armhf snap as well.

```
josepht@maple:~/snaps/joe$ ~/src/git/snapcraft/bin/snapcraft push ~/Downloads/hello-joe_0.4_armhf.snap --release edge
Pushing '/home/josepht/Downloads/hello-joe_0.4_armhf.snap' to the store.
Uploading /home/josepht/Downloads/hello-joe_0.4_armhf.snap [                                                  ]   0%
Uploading /home/josepht/Downloads/hello-joe_0.4_armhf.snap [==================================================] 100%
Ready to release!|                                                                                                  
Revision 5 of 'hello-joe' created.
Arch    Track    Series    Channel    Version    Revision
armhf   latest   16        stable     -          -
                           candidate  -          -
                           beta       -          -
                           edge       0.4        5
josepht@maple:~/snaps/joe$ ~/src/git/snapcraft/bin/snapcraft release hello-joe 5 0.1/edge
Arch    Track    Series    Channel    Version    Revision
armhf   0.1      16        stable     -          -
                           candidate  -          -
                           beta       -          -
                           edge       0.4        5
The '0.1/edge' channel is now open.
```

File: integration_tests/snaps/arm/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: basic
```
Comment (sergiusens) [inline integration_tests/snaps/arm/snapcraft.yaml]: is this registered on the integration servers?

Comment (josepht) [inline integration_tests/snaps/arm/snapcraft.yaml]: I guess.  I copied it from 'integration_tests/snaps/basic/snapcraft.yaml'

Comment (josepht) [inline integration_tests/snaps/arm/snapcraft.yaml]: It's overwritten in the test cases.

File: integration_tests/test_store_release.py [hunk 2]
```diff
@@ -105,13 +113,20 @@ def test_release_with_login_multiarch(self):
         expected = r'.*amd64   0.1      16        stable     -          -.*'
         self.assertThat(output, MatchesRegex(expected, flags=re.DOTALL))
 
+    @unittest.skipUnless(
+        (os.getenv('TEST_SNAP_WITH_TRACKS', 'fake') ==
```
Comment (sergiusens) [inline integration_tests/test_store_release.py]: this should just be

```python
def _can_test_track():
    test_tracks = os.getenv('TEST_SNAP_WITH_TRACKS') == 'test-snapcraft-tracks'
    integration_user = os.getenv('TEST_USER_EMAIL') == 'u1test+snapcraft@canonical.com'

    return test_tracks and integration_user
```

and then

```python
@skipUnless(_can_test_track())
def ...
```

Please verify the integration test.

Comment (sergiusens) [inline integration_tests/test_store_release.py]: Please verify the integration test using the staging store and using the fake server

Comment (josepht) [inline integration_tests/test_store_release.py]: Fake server:

```
josepht@maple:~/src/git/snapcraft$ ./runtests.sh integration test_store_release.py
Code coverage for Python.  Use '__main__.py help' for help.
test_release_with_login (test_store_release.ReleaseTestCase)
test_store_release.ReleaseTestCase.test_release_with_login ... [sudo] password for josepht: 
ok
test_release_with_login_arm (test_store_release.ReleaseTestCase)
test_store_release.ReleaseTestCase.test_release_with_login_arm ... skipped 'Skip CI only test'
test_release_with_login_multiarch (test_store_release.ReleaseTestCase)
test_store_release.ReleaseTestCase.test_release_with_login_multiarch ... skipped 'Skip CI only test'
test_release_without_login (test_store_release.ReleaseTestCase)
test_store_release.ReleaseTestCase.test_release_without_login ... ok

----------------------------------------------------------------------
Ran 4 tests in 14.114s

OK (skipped=2)
Everything passed
```

**Pull request merged**